gunicorn/http/wsgi.py

```diff
@@ -130,6 +131,7 @@ def create(req, sock, client, server, cfg):
             continue
         elif hdr_name == "CONTENT-LENGTH":
             environ['CONTENT_LENGTH'] = hdr_value
+            environ['wsgi.input_terminated'] = False
```

Comment:
Is this necessary? The way I read the gist it seems like Gunicorn could safely always set `wsgi.input_terminated`.

Our `LengthReader` will not allow reading past `Content-Length` bytes, but it will also return zero bytes to signal EOF at that point and it will raise `IOError` if the `recv` fails due to client disconnect.

I think we can just set it always.

Reply:
@tilgovi my initial intention was to keep the default behaviour for it waiting for a proper way to test how the implementation is used on different platforms. But maybe we should though we should try to sponsor it as an update of the wsgi protocol since it would be really the default....

Reply:
It cannot be the default that applications and frameworks assume `wsgi.input_terminated`, because some servers might not limit the number of bytes that can be read to `Content-Length`.

Gunicorn always does, though, so I think it's easier to think about if we always set it.

- The application cannot read past `Content-Length`, if it exists
- The `Content-Length` header, if it exists, is passed through to the application
- The input stream behaves the same always: `IOError` on client disconnect and EOF on chunk termination or `Content-Length` bytes read.

I cannot see a reason to ever set it to `False`, but I may be missing something.

Reply:
@mitsuhiko if I may ping you here, does this sound right?

Reply:
@tilgovi i guess your assumption is right. I think at the end we have the choice between:

1. beeing conservative: forcing the client to use the whole PEP3333 when a `Content-Length` header is given
2. always set `wsgi.input_terminated` to **true**. which is a breaking change at least for flask based applications
3. consider an incremental path with an environment variable setting the default for the case where we get a `Content-Length` header


Thoughts?

Reply:
It is breaking because it changes what the environment says, but it doesn't change behavior, does it? Maybe werkzeug would rely on Gunicorn to do what it was already doing itself?

Reply:
@tilgovi well it changes the behaviour for werkzeug and other applications/frameworks that handle `wsgi.input_terminated` hen they process a request. For gunicorn itself, this won't change much the way we are processing the request indeed.

Anyway I think you convinced me, I will make the change to set `wsgi.input_terminated` to true by default.

Reply:
What I mean is this.

Before: Werkzeug ensures that the stream is terminated, either on `Content-Length` bytes, last chunk, or through `IOError`.
After: Gunicorn ensures that the stream is terminated, either on `Content-Length` bytes, last chunk, or through `IOError`.

So, it's breaking because Werkzeug will work differently under Gunicorn, but it's _not_ breaking because the for the application the behavior is the same, it's just Gunicorn that provides the behavior rather than Werkzeug. And really, Gunicorn was already providing this behavior, just not advertising it with `wsgi.input_terminated`, right?

Reply:
We can ship this now as is. If we want to change it to always set `wsgi.input_terminated` to `True`, we have one more task that we can do in a separate ticket. We have to raise `IOError` if we detect a disconnect before `CONTENT_LENGTH` bytes.